### ISSUE TITLE
Istio setup and best practices

### ISSUE THREAD

**Issue by christian-roggia** [ory/oathkeeper#624]
This ticket addresses the setup with Istio + Envoy in a Kubernetes environment. This ticket provides a working example of an Istio setup and could be potentially adapted to any Envoy-based gateway such as Ambassador.

The version of Istio used is the latest available: 1.8.1.

Related ticket https://github.com/ory/oathkeeper/issues/263.

The following `EnvoyFilter` is working with a setup of ory Hydra (1.8.5) + Kratos (0.5.5) + Oathkeeper (0.38.5).

**I would like to request a review by the ory team and ask whether they could verify that all the headers used by oathkeeper have been whitelisted.**

Hopefully, this YAML will be useful and save some time and many headaches for all those who need to set up Istio as an edge proxy.

Our setup is working on both GKE and AWS with 2 ingresses (one secured with IAP and one public) which points to 2 Istio ingress gateways (NodePort) which authenticate and authorize requests via oathkeeper (EnvoyFilter).

```yaml
apiVersion: networking.istio.io/v1alpha3
kind: EnvoyFilter
metadata:
  name: oathkeeper
  namespace: istio-system
spec:
  workloadSelector:
    labels:
      istio: ingressgateway
  configPatches:
    - applyTo: HTTP_FILTER
      match:
        context: GATEWAY
        listener:
          filterChain:
            filter:
              name: "envoy.filters.network.http_connection_manager"
              subFilter:
                name: "envoy.filters.http.router"
      patch:
        operation: INSERT_BEFORE
        value:
          name: envoy.filters.http.ext_authz
          typed_config:
            '@type': type.googleapis.com/envoy.extensions.filters.http.ext_authz.v3.ExtAuthz
            failure_mode_allow: false
            http_service:
              path_prefix: /decisions
              server_uri:
                uri: http://oathkeeper-api.oathkeeper.svc.cluster.local:4456
                cluster: outbound|4456||oathkeeper-api.oathkeeper.svc.cluster.local
                timeout: 10s
              authorization_request:
                allowed_headers:
                  patterns:
                  - exact: accept
                  - exact: authorization
                  - exact: cookie
                  - exact: content-type
                  - exact: x-forwarded-for
                  - exact: x-forwarded-proto
                  - exact: x-forwarded-host
              authorization_response:
                allowed_upstream_headers:
                  patterns:
                  - exact: authorization
```

**Comment by christian-roggia** [comment]
This ticket could be (and probably should be) merged into the documentation.

**Comment by aeneasr** [comment]
Would absolutely love having that in the docs, maybe as a guide?

**Comment by sebastianvoss** [comment]
Thanks @christian-roggia, this was really helpful to set it up in Istio.

Is there also a way to set something like `X-User` header based on the identity returned by `oathkeeper`?

**Comment by christian-roggia** [comment]
> Thanks @christian-roggia, this was really helpful to set it up in Istio.
> 
> Is there also a way to set something like `X-User` header based on the identity returned by `oathkeeper`?

I believe you only need to whitelist the headers you would like to forward to the upstream service.

You can do so by adding the headers you need to the `allowed_upstream_headers` list. I think that in oathkeeper you can set them through a `mutator` rule.

**Comment by sebastianvoss** [comment]
@christian-roggia, you are right. It works a you described it. Thanks a lot!

**Comment by rauanmayemir** [comment]
Do you need to allow extra headers for tracing to work? I want oathkeeper filter request to be included as a span, i.e a part of the main request trace.

**Comment by christian-roggia** [comment]
I haven't tried it yet, but you can enable tracing on istio itself and probably you will need to whitelist some headers, I guess it also depends on the tracing platform you are using.. let me know if you manage to do it!

**Comment by rauanmayemir** [comment]
@christian-roggia I added tracing headers in `authorization_request.allowed_headers`, but I'm not sure. I think it works as it is without adding anything there. I will know more later.

**Comment by rajsenthil** [comment]
Thanks for the post @christian-roggia. I am looking for [similar setup](https://app.slack.com/client/T010GP8ELBT/C012RJ2MQ1H/thread/C012RJ2MQ1H-1619633871.437200) and this post provide a solution. 
Can you please confirm, for EnvoyFilter to work, whether ory Hydra is required here? 
Can the Envoy filter work with oathkeeper + kratos to intercept the request for a valid session cookie exists or not? 
Thanks again.

**Comment by christian-roggia** [comment]
I confirm that ory/hydra isn't required for the EnvoyFilter to work. You only need ory/oathkeeper as it provides the decision endpoint, you could technically deploy it alone with Istio/Envoy and nothing else in the cluster.

TL;DR: Kratos + Oathkeeper with Istio and this EnvoyFilter work just fine.

**Comment by batistein** [comment]
@christian-roggia Thank you very much for your example!! Do you know how to reduce the scope of the filter so it's not coupled on the gateway? We have a lot of other services running under the same gateway, where the oathkeeper shouldn't be used... I know this is an istio thing but I didn't found anything related to this topic. The only thing I saw was the newly introduced External Authorization from istio... as I understand the above example could be translated to this [object](https://istio.io/latest/docs/tasks/security/authorization/authz-custom/) as well right?

**Comment by christian-roggia** [comment]
You can use the `workloadSelector` to target only specific Pods, you just need to use your own labels instead of the labels of the ingressgateway

**Comment by batistein** [comment]
Thanks for the quick response @christian-roggia I tried it with a custom label which is on every pod I want to secure. Unfortunately, it seems like that the oathkeeper is not authorizing and bypassed. Do I also need to change the spec.configPatches.match.context=Gateway to something else? Thank you very much!

**Comment by christian-roggia** [comment]
Yes, you are right. I believe that changing to SIDECAR_INBOUND could be a step in the right direction?

**Comment by adamstrawson** [comment]
@christian-roggia Have you had any success defining oathkeeper  rules using`upstream` with `strip_path` alongside EnvoyFilters?

**Comment by nickjn92** [comment]
Later versions of Istio now offers a easier way to plugin external authorization: https://istio.io/latest/docs/tasks/security/authorization/authz-custom/
